src/api/database-operations/getItems.js

```diff
@@ -1,48 +1,110 @@
 const isDistinctExcluded = require('./../shared').isDistinctExcluded
 const _ = require('lodash')
 
+const APIMethod = require('../../APIMethod')
+const getQueryBuilder = require('../../buildQuery')
+const cloneMongooseQuery = require('../shared').cloneMongooseQuery
+
+const getCount = require('./getCount')
+
 /**
- * Given a model, erm options, and a list of excluded keys for model descendants, returns
- * Express middleware that builds a mongoose query to retrieve _all_ items that match the query,
- * executes the query, and stores the result (an array) in the request.
+ * Given global query options, a Mongoose context (just a ModelQuery), and a Mongo
+ * query, get all of the documents matching the query in the context.
+ *
+ * @param {Object} queryOptions - Global options to apply to all queries
+ * @param queryOptions.lean
+ * @param queryOptions.readPreference
+ * @param queryOptions.limit
+ *
+ * @param {ModelQuery} mongooseContext - The documents to query
+ * @param {Object} query - MongoDB query object to apply to the context
  *
- * @param {Object} model
- * @param {Object} options
- * @param {Object} excludedMap
- * @return {function(*=, *=, *=)}
+ * @return {Promise}
  */
-function getItems (model, options, excludedMap) {
-  const buildQuery = require('../../buildQuery')(options)
-  const errorHandler = require('../../errorHandler')(options)
-
-  return (req, res, next) => {
-    if (isDistinctExcluded(options.filter, excludedMap, req)) {
-      req.erm.result = []
-      req.erm.statusCode = 200
-      return next()
-    }
+function getItems (queryOptions, mongooseContext, query) {
+  const buildQuery = getQueryBuilder(queryOptions)
 
-    options.contextFilter(model, req, (filteredContext) => {
-      buildQuery(filteredContext.find(), req._ermQueryOptions).then((items) => {
-        req.erm.result = items
-        req.erm.statusCode = 200
-
-        if (options.totalCountHeader && !req._ermQueryOptions['distinct']) {
-          options.contextFilter(model, req, (countFilteredContext) => {
```

Comment:
In the old implementation, `contextFilter()` was called twice here -- once with the original request passed to the middleware, and then again with a _mutated_ request that has `items` attached and the status code set.

It is _theoretically_ possible that someone's `contextFilter()` depended on this behavior for this middleware (but I don't think it's likely). 

The new operations are "atomic" in the sense that they don't mutate the request until the middleware has finished running.

These changes still pass all of the tests we've got.